src/constructors/mycielskian.c

```diff
@@ -0,0 +1,245 @@
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "igraph_constructors.h"
+#include "igraph_operators.h"
+
+#include "igraph_conversion.h"
+#include "igraph_interface.h"
+
+#include "core/interruption.h"
+#include "math/safe_intop.h"
+
+/**
+ * \function igraph_mycielskian
+ * \brief Generate the Mycielskian of a graph with \p k iterations.
+ *
+ * The Mycielskian construction is a method used to generate a triangle-free
+ * graph with an increased chromatic number. Given an input graph, the
+ * Mycielskian transformation produces a new graph where the chromatic number
+ * increases by at most one in each iteration.
+ *
+ * </para><para>
+ * This transformation is useful in graph theory for constructing
+ * graphs with large chromatic numbers while avoiding small cycles,
+ * particularly triangles.
+ *
+ * </para><para>
+ * Let the \c n vertices of the given graph G be \c v_1, ..., \c v_n.
+ * The Mycielski graph \c M(G) contains G itself as a subgraph,
+ * together with \c n+1 additional vertices:
+ * 
+ * \olist
+ * \oli A vertex \c u_i corresponding to each vertex \c v_i of \c G.
+ * \oli An extra vertex \c w.
+ * \endolist
+ * 
+ * </para><para>
+ * The edges are added as follows:
+ * \olist
+ * \oli Each vertex \c u_i is connected to \c w, forming a star.
+ * \oli For each edge <code>(v_i, v_j)</code> in \c G, two new edges are added:
+ *   <code>(u_i, v_j)</code> and <code>(v_i, u_j)</code>.
+ * \endolist
+ * 
+ * Thus, if G has \c n vertices and \c m edges, the Mycielski graph \c M(G) 
+ * has <code>2n + 1</code> vertices, and <code>3m + n</code> edges.
+ *
+ * </para><para>
+ * The k-th iterated Mycielskian has:
+ * <code> n_k = (n + 1) * 2^k - 1 </code>
+ * vertices, where \c n is the number of vertices in the original graph.
+ *
+ * The edge count increases as:
+ * <code> m_k = ((2m + 2n + 1) * 3^k - n_{k+1}) / 2 </code>
+ * 
+ * where m is the original number of edges. The number of edges increases
+ * from \c m to <code>3m + n</code> in each iteration, as we add 2 edges for every
+ * existing edge and 1 edge for every vertex.
+ *
+ * </para><para>
+ * The Mycielskian is commonly used to demonstrate that graphs can have high
+ * chromatic numbers without forming small cycles (i.e., triangle-free graphs).
+ * The Grotzsch graph, for example, is obtained as \c M_4 of a triangle-free graph.
+ * </para><para>
+ * 
+ * \param graph Pointer to the input graph.
+ * \param res Pointer to an uninitialized graph object where the Mycielskian
+ *        of the input graph will be stored.
+ * \param k Integer, the number of Mycielskian iterations (must be non-negative).
+ * \return Error code.
+ *
+ * \sa \ref igraph_mycielski_graph().
+ *
+ * Time complexity: Exponential in \p k.
+ */
```

Comment:
Something important is missing: a discussion of the special case for the null graph and singleton graphs.

We also must mention the behaviour with directed graphs, non-simple graphs and disconnected graphs.